extension/src/lttb.rs

```diff
@@ -0,0 +1,219 @@
+
+use pgx::*;
+use std::borrow::Cow;
+
+use crate::{
+    aggregate_utils::in_aggregate_context, flatten, json_inout_funcs, palloc::Internal, pg_type,
+};
+
+use time_weighted_average::tspoint::TSPoint;
+
+use flat_serialize::*;
+
+// hack to allow us to qualify names with "timescale_analytics_experimental"
+// so that pgx generates the correct SQL
+mod timescale_analytics_experimental {
+    pub(crate) use super::*;
+    extension_sql!(r#"
+        CREATE SCHEMA IF NOT EXISTS timescale_analytics_experimental;
+    "#);
+}
+
+extension_sql!(r#"
+CREATE TYPE timescale_analytics_experimental.SortedTimeseries;
+"#);
+
+pg_type! {
+    #[derive(Debug)]
+    struct SortedTimeseries {
+        num_points: u64,  // required to be aligned
```

Comment:
Would you be losing anything by making this a general Timeseries object with a 32 bit size and a bit indicating that it's sorted?

Reply:
🤔 we'd have to do something like
```rust
SortedTimeseries {
    num_points: u32,
    sorted: bool,
    padding: [u8; 3],
    points: [TsPoint; self.num_points],
}
```
but it _should_ work.

That said, I'm not sure if it's all that useful; for one, I'm not sure yet if we ever want to return an unsorted timeseries back to the user, for another, I think knowing that a timeseries is sorted at the type level is useful, it means you don't have to do a defensive re-sort if you want sorted output after `unnest`. Since there's no real way to do a data-dependent sort in SQL, I think storing it internally would force the user to always sort the output of LTTB. (Another option is to add a `sort_timeseries()` function, but that might force us to re-allocate the timeseries even if it's already sorted, I'm not confident enough about the expression context to say whether it is or isn't required)